### ISSUE TITLE
twinx changes figure size unexpectedly when used with inset axes.

### ISSUE THREAD

**Issue by eshanking** [matplotlib/matplotlib#20607]
### Bug report

When using twinx in the following setting I find that it dramatically changes the figure size, sometimes exceeding the 2^16 pixel limit.

**Code for reproduction**

```python
import matplotlib.pyplot as plt
import numpy as np

fig,ax = plt.subplots(figsize=(5,5))

# create a horizontal bar chart
barx = [1,2,3]
bary = np.array([1,2,3])/3

rects = ax.barh(barx,bary)

# create some data to plot
x = np.arange(5000)

insets = [] # list to hold inset axes

# create inset axes at each level of the bar chart
for num in range(len(barx)):
    ypos = rects.patches[num].get_y()
    xpos = 1.2
    
    ax1 = ax.inset_axes([xpos,ypos,0.4,0.8],transform=ax.transData)
    
    ax1.plot(x,x**num,linewidth=2)
    insets.append(ax1)

# create twin axes for each inset axes
twins = []
for a in insets:
    twin_ax = a.twinx()
    twins.append(twin_ax)
```

**Actual outcome**

<img width="414" alt="image" src="https://user-images.githubusercontent.com/33136282/124974360-b36cc480-dffa-11eb-992f-519de08db37d.png">

**Expected outcome**

This is what the figure looks like without using twinx():

<img width="1196" alt="image" src="https://user-images.githubusercontent.com/33136282/124974475-d7300a80-dffa-11eb-80c7-43c5539cb69a.png">

**Matplotlib version**

  * Operating system: macOS Big Sur 11.4
  * Matplotlib version (`import matplotlib; print(matplotlib.__version__)`): 3.3.4
  * Matplotlib backend (`print(matplotlib.get_backend())`): 'module://matplotlib_inline.backend_inline'
  * Python version: 3.9.5

I installed matplotlib and python with conda.

**Comment by jklymak** [comment]
What happens if you do twin_ax.set_in_layout(False) ?

**Comment by tacaswell** [comment]
This is definitely related to `bbox_inches='tight'` as 

![so](https://user-images.githubusercontent.com/199813/124975581-29256000-dffc-11eb-828a-705218dc03eb.png)

is the result of doing `fig.savefig(...)`

I think that 

```python
import matplotlib.pyplot as plt

fig, axd = plt.subplot_mosaic(
    "AB\nAC\nAD", gridspec_kw={"width_ratios": [3, 1]}, constrained_layout=True
)
for k in "BCD":
    axd[k].twinx()
# plt.show()

```

![so2](https://user-images.githubusercontent.com/199813/124976549-6dfdc680-dffd-11eb-8f52-856b16ea6e14.png)

will get what you want in a slightly more direct way?

**Comment by eshanking** [comment]
> What happens if you do twin_ax.set_in_layout(False) ?
Thank you for the reply!

I tried:
```python
for a in insets:
    twin_ax = a.twinx()
    twin_ax.set_in_layout(False)
    twins.append(twin_ax)
```

Which did not seem to have an effect.

**Comment by eshanking** [comment]
> This is definitely related to `bbox_inches='tight'` as
> 
> ![so](https://user-images.githubusercontent.com/199813/124975581-29256000-dffc-11eb-828a-705218dc03eb.png)
> 
> is the result of doing `fig.savefig(...)`
> 
> I think that
> 
> ```python
> import matplotlib.pyplot as plt
> 
> fig, axd = plt.subplot_mosaic(
>     "AB\nAC\nAD", gridspec_kw={"width_ratios": [3, 1]}, constrained_layout=True
> )
> for k in "BCD":
>     axd[k].twinx()
> # plt.show()
> ```
> 
> ![so2](https://user-images.githubusercontent.com/199813/124976549-6dfdc680-dffd-11eb-8f52-856b16ea6e14.png)
> 
> will get what you want in a slightly more direct way?

Thanks for the reply! 

I did originally use gridspec, but I would love to do this by positioning the inset axes explicitly by the position of the bar chart rectangles. If not, the axes don't always end up exactly aligned. Another option is to convert the rectangle positions to figure coordinates using transData somehow, but I figured that using inset_axes would avoid this headache.

**Comment by jklymak** [comment]
The bbox is not being properly updated with the twin on there.  Maybe its fixable.  The workaround is to call an extra draw before you display:

putting `fig.draw_no_output()` fixes the problem when saving to png with `bbox_inches='tight'`

**Comment by jklymak** [comment]
Also if you do not need twins, but just want a different axes on the right hand side, secondary_yaxis is less fragile. 

```python 
for a in insets:
    twin_ax = a.secondary_yaxis('right', functions=(lambda x: x, lambda x: x))
    twins.append(twin_ax)
```

**Comment by jklymak** [comment]
I can reproduce on 3.3.4, but this is fixed on 3.4.2.  Please upgrade matplotlib.
